## Problem #92: Square Digit Chains
[Link to Problem](https://projecteuler.net/problem=92)

### Problem Description

A number chain is created by continuously adding the square of the digits in a number to form a new number until it has been seen before.

For example,

$$

44 \to 32 \to 13 \to 10 \to \mathbf 1 \to \mathbf 1 \\[5pt]
85 \to \mathbf{89} \to 145 \to 42 \to 20 \to 4 \to 16 \to 37 \to 58 \to \mathbf{89}

$$

Therefore any chain that arrives at $1$ or $89$ will become stuck in an endless loop. What is most amazing is that EVERY starting number will eventually arrive at $1$ or $89$.

How many starting numbers below ten million will arrive at $89$?

### Approach

We can try progressively find at which number the chain ends for each number using a DFS-like approach.

In [6]:
L = 10_000_000

tail = {1: 1, 89: 89}

def dfs(x):
    if x in tail:
        return tail[x]
    s, y = 0, x
    while y:
        s += (y % 10) ** 2
        y //= 10
    tail[x] = dfs(s)
    return tail[x]

for i in range(1, L):
    dfs(i)

count = 0

for i in range(1, L):
    if tail[i] == 89:
        count += 1

print(count)

8581146


###### Result: **8581146** | Execution time: ~17s

### Complexity analysis

Time complexity: $O(L)$, as each number goes in the function once

##### Tags: #cycles, #implementation, #dfs